### PR TITLE
feat: add append_key_value_metadata

### LINKED ISSUE #3356: Is it possible for ArrowWriter to write key_value_metadata after write all records
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The metadata of parquet is placed in the footer, In this design, metadata can be written after the data has been written, which allow for single pass writing.

But in current implementation of ArrowWriter, we can only provide WriterProperties in `try_new`, for most fields of WriterProperties this is fine, but not the case for key_value_metadata.

key_value_metadata is meant for users to save customized information, for example we can encode some stats in here.

This usually means that we need to loop the original data to get stats information, this means we have to loop data twice:
1. First to get the stats information
2. Second to write to parquet

If ArrowWriter could provide a way to set key_value_metadata  after write all records, then only one loop is need.

**Describe the solution you'd like**

Add new method in ArrowWriter/SerializedFileWriter to let users set key_value_metadata.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jiacai2050**
# Which issue does this PR close?

Closes #3356.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

Yes, add a new method in ArrowWriter: `append_key_value_metadata`

### CODE REVIEW

**Comment by tustvold** [comment]
Perhaps we could make this append, as opposed to replacing the key value metadata in WriterProperties?

**Comment by jiacai2050** [comment]
Make sense, will address this soon.

**Review by tustvold — APPROVED** [review]
Just some minor nits, thank you

File: parquet/src/file/writer.rs [hunk 1]
```diff
@@ -298,7 +301,11 @@ impl<W: Write> SerializedFileWriter<W> {
             row_groups,
             version: self.props.writer_version().as_num(),
             schema: types::to_thrift(self.schema.as_ref())?,
-            key_value_metadata: self.props.key_value_metadata().cloned(),
+            key_value_metadata: self.props.key_value_metadata().cloned().map(|kv| {
```
Comment (tustvold) [inline parquet/src/file/writer.rs:304]: This won't work correctly if `props.key_value_metadata` isn't present... Which also makes me realise this needs a test. I'll bash something out quickly

Comment (jiacai2050) [reply parquet/src/file/writer.rs]: 👍 Good catch

**Pull request merged**